### PR TITLE
Add note about alertmanager config encoding

### PR DESCRIPTION
**Pull request by leifmadsen**
Add a node and example about how to deal with complicated
alertmanager.yaml configurations. Suggests encoding the
alertmanager.yaml configuration as base64 and using the data definition
method in the Secret rather than stringData and defining the
configuration in plaintext.

Related: rhbz#2023763
Signed-off-by: Leif Madsen <lmadsen@redhat.com>

### CODE REVIEW

**Comment by leifmadsen** [comment]
Joanne, I wasn't exactly sure whether a note here at the top was appropriate, whether I should just not wrap it in a note, or whether I should add the note maybe somewhere else, and just add this as an alternative procedure.

We could also just make this the default and preferred definition method, but it does result in an extra step being required that might not otherwise be required.

Happy to hear some thoughts on how to construct this. I could probably also create a clever `oc patch` command that would allow someone to pass their configuration as I have through `base64` and save it to an environment variable (`ALERTMANAGER_YAML=$(cat /tmp/template.jinja2 | base64 --wrap=0`) and the create the `Secret` template as we do now, pass it to `oc patch` and include the contents via the environment variable. The issue I have is it might look a bit awkward once it's all together and ultimately just look too clever? I'm not really sure here.

**Comment by leifmadsen** [comment]
Per meeting today, here are next steps:

* note will be converted to a new procedure, called "Creating an alert route with templating in Alertmanager"
* will be similar procedure as 5.3.3 with content of note added as steps
* TODO: Leif to provide the content of the manifest, Joanne to scaffold the file. Leif will use a simplified examples from that provided by the original BZ report.
* update 5.3 for the TOC and referencing both procedures, "Alert routes can be created in one of two methods".

**Review by pleimer — APPROVED** [review]

**Comment by csibbitt** [comment]
I just read this and I find the instructions clear enough, but normally I think we could use an example manifest to go along with something like this. Just a trivial alertmanager config with a single key should suffice

**Comment by leifmadsen** [comment]
> I just read this and I find the instructions clear enough, but normally I think we could use an example manifest to go along with something like this. Just a trivial alertmanager config with a single key should suffice

That's exactly what I'm adding now :)

File: doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-in-alertmanager.adoc [hunk 1]
```diff
@@ -1,25 +1,24 @@
 [id="creating-an-alert-route-in-alertmanager_{context}"]
 = Creating an alert route in Alertmanager
 
-[role="_abstract"]
-Use Alertmanager to deliver alerts to an external system, such as email, IRC, or other notification channel. The Prometheus Operator manages the Alertmanager configuration as a {OpenShift} secret. By default, {Project} ({ProjectShort}) deploys a basic configuration that results in no receivers:
+include::../modules/con_alertmanager-route.adoc[]
```
Comment (csibbitt) [inline doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-in-alertmanager.adoc:4]: I couldn't find another example of a concept being linked from a procedure like this. Should this move above line 39 in the assembly?

File: doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-with-templating-in-alertmanager.adoc [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+[source,bash]
+----
+$ oc run curl --image=radial/busyboxplus:curl -i --tty
+----
+
+. Run `curl` against the `alertmanager-operated` service to retrieve the status and `configYAML` contents, and verify that the supplied configuration matches the configuration in Alertmanager:
++
+[source,bash,options="nowrap"]
+----
+[ root@curl:/ ]$ curl alertmanager-operated:9093/api/v1/status
```
Comment (csibbitt) [inline doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-with-templating-in-alertmanager.adoc:91]: Does this work? It's not supposed to:

https://github.com/infrawatch/service-telemetry-operator/blob/master/roles/servicetelemetry/templates/manifest_alertmanager.j2#L14

File: doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-in-alertmanager.adoc (hunk 1 above)
Comment (leifmadsen) [reply doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-in-alertmanager.adoc:4]: I might consider naming this something else. It's not really a "concept" as much as it's part of a de-duplication effort. It felt kinda like a concept, but it's not a top level concept like the other modules.

Maybe I should just use `proc_creating-...-common-intro.adoc` instead.

File: doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-with-templating-in-alertmanager.adoc (hunk 2 above)
Comment (leifmadsen) [reply doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-with-templating-in-alertmanager.adoc]: It does for STF 1.3, but not STF 1.4. I want to backport all of these changes to STF 1.3, so I'm ignoring that it is invalid for STF 1.4. I have another task where I'm going to update all of this after this set of changes merges and is backported. See https://issues.redhat.com/browse/STF-913

Comment (csibbitt) [reply doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-with-templating-in-alertmanager.adoc]: https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#alertmanagerspec

I don't happen to have a cluster up at the moment to check if this is still working - it was when I wrote the listenLocal line

**Comment by leifmadsen** [comment]
Ready for another round of reviews. I will probably come back and adjust the `con_...` file I added to be `proc_creating-...-common-intro.adoc` though.

File: doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-with-templating-in-alertmanager.adoc (hunk 2 above)
Comment (csibbitt) [reply doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-with-templating-in-alertmanager.adoc]: Okay, that makes sense. Thanks.

**Review by csibbitt — APPROVED** [review]

**Review by pleimer — APPROVED** [review]
LGTM

**Comment by leifmadsen** [comment]
Once this merges, it will need to be backported to STF 1.3, but then additional work will be required here to bring this up to STF 1.4 implementation which has stricter security requirements / interfaces.

File: doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-common-procedure.adoc [hunk 3]
```diff
@@ -0,0 +1,38 @@
+. To verify the configuration is loaded into Alertmanager, create a pod with access to `curl`:
```
Comment (JoanneOFlynn2018) [inline doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-common-procedure.adoc:1]: When you say access to curl here, what do you mean? Is this the curl pod mentioned down further? I only know curl as a CLI command but maybe techy users will know what you mean immediately.

File: doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-common-procedure.adoc [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
++
+[source,bash]
+----
+$ oc run curl --image=radial/busyboxplus:curl -i --tty
+----
+
+
+. Run `curl` against the `alertmanager-operated` service to retrieve the status and `configYAML` contents, and verify that the supplied configuration matches the configuration in Alertmanager:
```
Comment (JoanneOFlynn2018) [inline doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-common-procedure.adoc:9]:
```suggestion
. Run `curl` against the `alertmanager-operated` service to retrieve the status and `configYAML` contents, and verify that the supplied configuration matches the configuration in Alertmanager:
```
```suggestion
. Run the `curl` command against the `alertmanager-operated` service to retrieve the status and `configYAML` contents, and verify that the supplied configuration matches the configuration in Alertmanager:
```

File: doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-in-alertmanager.adoc [hunk 5]
```diff
@@ -77,7 +82,7 @@ spec:
         - name: 'null'
 ----
 
-. Verify that the configuration is applied to the secret:
+. Verify that the configuration has been applied to the Secret:
```
Comment (JoanneOFlynn2018) [inline doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-in-alertmanager.adoc:85]:
```suggestion
. Verify that the configuration has been applied to the secret:
```

File: doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-with-templating-in-alertmanager.adoc [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+----
+
+. Edit the `ServiceTelemetry` object for your {ProjectShort} deployment:
++
+[source,bash]
+----
+$ oc edit stf default
+----
+
+. To deploy a custom Alertmanager route with {ProjectShort}, you must pass an `alertmanagerConfigManifest` parameter to the Service Telemetry Operator that results in an updated Secret that is managed by the Prometheus Operator:
```
Comment (JoanneOFlynn2018) [inline doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-with-templating-in-alertmanager.adoc:46]:
```suggestion
. To deploy a custom Alertmanager route with {ProjectShort}, you must pass an `alertmanagerConfigManifest` parameter to the Service Telemetry Operator that results in an updated secret that is managed by the Prometheus Operator:
```

Comment (JoanneOFlynn2018) [inline doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-with-templating-in-alertmanager.adoc]:
```suggestion
. To deploy a custom Alertmanager route with {ProjectShort}, you must pass an `alertmanagerConfigManifest` parameter to the Service Telemetry Operator that results in an updated secret that is managed by the Prometheus Operator:
```

File: doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-common-procedure.adoc (hunk 3 above)
Comment (leifmadsen) [reply doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-common-procedure.adoc:1]: I really just mean launch a pod that has the `curl` binary available / installed.

Comment (JoanneOFlynn2018) [reply doc-Service-Telemetry-Framework/modules/proc_creating-an-alert-route-common-procedure.adoc:1]: OK, I get it!

**Review by JoanneOFlynn2018 — APPROVED** [review]

**Pull request merged**